### ISSUE TITLE
Unit tests for data import / export

### ISSUE THREAD

**Issue by jestockda** [dward2/BME547#518]
@dward2 

The prompt for the cell analysis assignment has the following expectation: 

"Unit testing exists for all functions that do any algorithmic work, excluding input/output routines"

Typically I think of input / output as accepting keyboard input or printing texting to the display, but in this case, do input / output routines include functions that exclusively handle data import / export? To be clear, not parsing incoming text nor packaging the data for export — just the very narrow scope of reading / writing to a file. Or is this sort of task something that would be acceptable to live within a script _outside_ of a function (given the functions exist on their own already). 

Thanks!

**Comment by dward2** [comment]
You bring up a good distinction.  For `pytest` it is problematic to test a function that requires keyboard or any other kind of interactive input.  But, a function that takes input in the form of reading in a file can be tested in `pytest.

However, if a function only uses existing Python commands or functions from published packages that ONLY read in data from a file and doesn't need to do any kind of manipulation, then it doesn't need to be tested.  But, if that input makes any kind of decisions about what data to import or not does require testing.

For example:
```python
def input_file(filename):
    with open(filename, 'r') as in_file:
        all_data = in_file.readlines()
    return all_data
```
The above function does not need a unit test because it uses a standard Python function to only input data.  It doesn't do anything else.  Another example:
```python
def output_json(my_dictionary):
    filename = "{}-{}.txt".format(my_dictionary.patient, datetime.now())
    with open(filename, 'w') as out_file:
        json.dump(my_dictionary, out_file)
```
The function above does need a unit test because it is doing data manipulation when it is creating the filename.  That is algorithmic work you added and needs to be tested.  So, I would either remove the creation of the filename to another testable function, or I would write a unit test for this function that checks that the appropriately named output file was created.

So, in short, if the function literally only reads in a file or outputs a file with no decisions being made, it doesn't need a unit test.

**Comment by jestockda** [comment]
For example, I imagine a script like this:

```
in_file = open("input_file.txt", "r")
all_lines = in_file.readlines()
in_file.close()

*call custom functions with unit testing*

out_file = open("filename.json", "w")
json.dump(*outputs generated by custom functions*)
out_file.close()

```

I don't see a reason to create 'wrapper' functions for these tasks, but want to adhere to guidelines / meet expectations.

**Comment by jestockda** [comment]
Thanks! Sorry for the redundant / overlapping message. Seems like we're largely on the same page, outside of the role of scripting vs. wrapper functions. I assume it's fine to just call existing functions directly?

**Comment by dward2** [comment]
If these lines of code are not wrapped in a function, then they will be executed when you import this module for use in `pytest` or another program when you may not want them executed.

So, at a minimum, they should be inside the `if __name__ == "__main__":` clause.

An advantage of putting them in a function is that they can be more easily reused.  But, if you don't need them reused, then I agree they could be part of the script.

**Comment by jestockda** [comment]
Ah. I'm planning a module + separate script that utilizes said module, but you make good points :)

Thanks for responding so quickly. Very helpful.
